In [1]:
import pandas as pd

df = pd.read_csv("messy_sales.csv")

print(df.shape)
display(df.head(10))
print(df.dtypes)

(300, 6)


,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id      int64
date            str
product         str
price       float64
qty           int64
zip           int64
dtype: object


Initial Observations:
-Some values in the `price` column are missing.
-The `zip` column is stored as a number, which can remove leading zeros.
-The `date` column is stored as text instead of a datetime type.

In [2]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [3]:
median_price = df["price"].median()

print("Median price:", median_price)

df["price"] = df["price"].fillna(median_price)

df.isna().sum()

Median price: 37.53


order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [4]:
print("Duplicate rows:", df.duplicated().sum())

df = df.drop_duplicates()

print("New shape:", df.shape)

Duplicate rows: 8
New shape: (292, 6)


In [5]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

df[["zip"]].head(10)

,zip
0,60614
1,30303
2,10001
3,98101
4,90405
5,30303
6,02134
7,90405
8,60614
9,90210


In [6]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

print(df.dtypes)

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object


In [7]:
print("Shape:", df.shape)
print("\nMissing values:")
print(df.isna().sum())
print("\nZIP lengths:")
print(df["zip"].str.len().value_counts())

Shape: (292, 6)

Missing values:
order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

ZIP lengths:
zip
5    292
Name: count, dtype: int64


In [8]:
df[df["qty"] < 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


In this, I saw that 3 rows had negative qualities, so I decided to keep them because they may represent product returns opposed to data-entry errors. Basically, the removal or change them without more information could make the sales data inaccurate.

In [9]:
df.to_csv("sales_clean.csv", index=False)

CLeaning LOG

I started with 300 rows and 6 columns.

-There were 12 missing prices, so I filled them in with the median price of $37.53. I used the median because really high or low prices would affect it less.

-I found 8 duplicate rows and removed them, which left 292 rows.

-I fixed the ZIP codes by changing them to text then adding back the leading zeros so that they were all 5 digits.

-I changed the date column into datetime format so all the dates were stored the same way.

-I found 3 rows with negative quantities. I kept them because they could just be returns and not actual mistakes.

If I were to clean the data incorrectly, it could mess up the sales numbers and make the business come to an incorrect conclusion.